model/outgoing_webhook.go

```diff
@@ -30,6 +33,7 @@ type OutgoingWebhook struct {
 	ContentType  string      `json:"content_type"`
 	Username     string      `json:"username"`
 	IconURL      string      `json:"icon_url"`
+	SecretToken  string      `json:"secret_token"`
```

Comment:
Can you elaborate on why incoming webhooks have a more elaborate schema here? It seems like we only support one kind of signing model for outgoing webhooks, but then allow variations for incoming? Should we simplify to just require a single signing model in all cases, even if that reduces flexibility?

Reply:
> It seems like we only support one kind of signing model for outgoing webhooks

That's right, we're using HMAC-SHA256.

> then allow variations for incoming

We currently support only two dominant models: HMAC-SHA1 and HMAC-SHA256. The second one is favored by many while Github uses HMAC-SHA1 without timestamp.

> Should we simplify to just require a single signing model in all cases, even if that reduces flexibility?

Even with a single model, there will be many variations and formats. For example, the custom headers, custom orders of the signed content, custom prefixes for the header values, whether timestamp or not. There seems no standard for even only HMAC-SHA1 signature. I think the major work for supporting signed incoming webhook is really the reconstruction of components used for signing. We need the reconstruction logic for a single model as much as multiple models.

Reply:
@tennom, thanks for the clarity. 

> There seems no standard for even only HMAC-SHA1 signature. I think the major work for supporting signed incoming webhook is really the reconstruction of components used for signing. We need the reconstruction logic for a single model as much as multiple models.

@jupenur & @tennom, wondering if we should just document the reconstruction logic and support exactly one, eliminating the need to model the others? Or do other products try to support multiple models this way?

Reply:
@lieut-data Honestly, I didn't have a chance to look at how others implemented webhook authentication. I came up with this approach and implemented it. Now I used the name "model" for modeling three signing components: signature header, timestamp header and signed content. By saying model, I mean splitting header strings, striping off prefixes and reconstructing the signed content with timestamp. However, here a model is not a separate authentication form, i.e. an alternative to HMAC. I needed the modeling because the names and formats of these signing components are different by webhook senders. If we eliminate modeling the signing components, we may need to save presets of these signing components for the supported webhook senders. For that we will need to split header strings, strip off prefixes and reconstruct the signed content with timestamp as we do in modeling if we support multiple webhook senders but I assume less modeling. 
Here modeling helps to support multiple webhook senders. Without modeling, the receiver may support only one sender or the ones in presets. The goal is how would the receiver automatically reconstruct the signing components given custom headers and custom format of signed content.  In the mean while, I will try to read how other's implemented. Please let me your thought, it may be major overhaul.

Reply:
Yup. This was very confusing to me as well when I started going through it, but I think what @tennom has done here makes a lot of sense. It's complicated, sure, but I don't see a better way around it, either.

> Or do other products try to support multiple models this way?

I don't think many other products support incoming webhooks like we're trying to here. Integrations tend to be more custom-built and single-purpose. The complexity comes from trying to be generic enough for all use-cases.

Reply:
@jupenur & @tennom, thanks for the detail here.

> Without modeling, the receiver may support only one sender or the ones in presets.

This is the part I'm working towards understanding. It sounds like we're trying to support the various types of senders and their existing expectations to sign requests. I was expecting we'd pick just one model and require senders to conform to that.

We already require a custom payload (i.e. the body of the webhook), so it doesn't seem a stretch to also require an explicit signing strategy. I don't think we're expecting to wire up a GitHub outgoing webhook directly to a Mattermost incoming webhook: they already need a translation layer. If the sender doesn't have built-in support for our approach, we'd encourage them to find a library/package that does, just like we require them to modify their payloads to meet our webhook schema.

The upside to supporting just one model is that our implementation becomes dramatically simpler, not just in code, but in testing effort: QA need only verify one path, instead of potentially several permutations.

Thoughts?

Reply:
@lieut-data & @jupenur 
I'm using phone to type, it's night here in Tibet, sorry.

1. in my research so far, imposing senders to conform to our signing strategy is not possible. Because the senders are not sending to MM only but multiple webhook receiver endpoints. for example, github webhooks have only one strategy for everybody, for Slack, MM and any others. Another way to prove this is MM will have only one signing strategy for outgoing webhooks for everybody, I.e. Slack and Stripe and all. If we needed to provide a signing strategy for each receiver, that would be more complicated than anything.
2. I'd quote your comment here but dont know how using a phone:) I'm not sure how we required a custom body from webhook senders. At least in the webhook authentication part, we don't assume or require any special format of the body or headers. We impose nothing to senders but we dont support headers in http authentication header, which seems not so popular, correct me if I got that wrong. 

3. Integrations in Github or others are done by repo owners, not the platforms developers,  if I'm not wrong. They usually have a choice of entering the preshared SecretToken and then the rest is automated by webhook sender platform. I think no translation layer will be created between Github and MM for signing, and the outgoing webhooks from Github are directly wired to MM as without the signing. Please share your ideas?

Reply:
@tennom, no worries re: typing on phone, and I hope I'm not bugging you at night! Feel free to respond whenever works best :)

I think the underlying assumption that needs to be vetted is whether or not we expect to wire up senders (like GitHub) /directly/ to Mattermost, or through an intermediate service.

My experience has been that because the two don't speak the same "payload language", an intermediate is always required. An outgoing webhook from Mattermost doesn't natively make sense to GitHub, unless some intermediate parses it and decides to take some action. Similarly, an incoming webhook from GitHub won't mean anything to Mattermost, unless some intermediate parses it and creates a post or some such.

My expectation is that this intermediate will be responsible for validating signatures. Given the need to "speak both languages", it will simply implement support for whatever each side supports. Thus, we don't need to worry about being generic here.

If, on the other hand, we have examples of directly wiring up senders to receivers, then it begins to make more sense to support multiple models: but it probably also means we should look at supporting the same set of models for the outgoing webhook for all the same reasons.

Do we know of any such examples of directly wiring up Mattermost to some service (that supports some kind of authentication scheme)?

Reply:
@lieut-data Thanks for all the explanation, now we have cleared out more.

I really didn't pay attention to normalization of the the payload and I didn't know that either. @jupenur you mentioned about normalization a lot, you might also talked about this?

Yes, I agree that if we don't have a normalized approach to consume payloads, having a normalized authentication process is useless. My generic way to normalize payloads would be letting the MM endpoint owner to enter payload json keys (start out with 3 keys for simplicity) and consume them in a predefined way. The drawback is more input from MM integraters but some intuitive UI design would help. For coding, it's simple as adding a field in struct and append them to `text.Props` or a predefined field.

Then this implementation would be the intermediate service that automates the incoming authentication and payload normalization for MM. If we have the intermediate service, this workflow makes senses because parse the payload only if authentication passed. If we use a intermediate service from third party, then we can get rid of the normalization and directly plug in the values and validate signatures as you said. 

> My expectation is that this intermediate will be responsible for validating signatures. 

If we have a third party as this intermediate service to validate the signatures, where should the `SecreToken` s be generated, on the third party platform or MM? 

> Do we know of any such examples of directly wiring up Mattermost to some service (that supports some kind of authentication scheme)?

No, I don't. Does MM already have an intermediate service to parse webhook payloads since MM takes in incoming webhooks? I just saw the code for supporting Slack's attachment in `App.HandleIncomingWebhook`.

 Let me list the possible next actions:

1. We get rid all the normalizing of signing components and validate on values directly, and expect an intermediate service to feed in the values. In this case, webhook authentication will not work without the intermediate.
2. We have an intermediate that normalizes payloads to MM'f form already. It would be easy to integrate the authentication as a part of it.
3. We start out with supporting signatures from Github and maybe a couple more explicitly. I will need to rework on code to do so.
4. We keep the current generic way for normalizing signing components and normalize payloads in this way.
5. What else?

Reply:
> I really didn't pay attention to normalization of the the payload and I didn't know that either. @jupenur you mentioned about normalization a lot, you might also talked about this?

What I was talking about was more closely related to validating signatures, whereas @lieut-data is now making a good point about the supported payload formats themselves. I got tangled up in the complexity of the change and totally missed that.

Supporting multiple types of signatures may still make sense if there are client implementations that have both explicit support for Slack (which we're compatible with) *and* a custom signature scheme (which Slack doesn't support for incoming webhooks). Otherwise I think the goal should be to simplify as much as possible.

Reply:
Thanks @jupenur and @lieut-data can you further direct me how we gonna handle this?

Reply:
Apologies for the delay here, @tennom! I think we've reached a point where it might make sense to hop on a call and discuss further. Let me ping you both on community.

Reply:
Conversation picking up over at https://community-daily.mattermost.com/core/pl/bbcek36f3jyepxnziykqwedpnc